perf/reader_test.go

```diff
@@ -0,0 +1,374 @@
+
+	"golang.org/x/sys/unix"
+)
+
+func TestPerfReader(t *testing.T) {
+	prog, events := mustOutputSamplesProg(t, 5)
+	defer prog.Close()
+	defer events.Close()
+
+	rd, err := NewReader(events, 4096)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer rd.Close()
+
+	ret, _, err := prog.Test(make([]byte, 14))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if errno := syscall.Errno(-int32(ret)); errno != 0 {
+		t.Fatal("Expected 0 as return value, got", errno)
+	}
+
+	record, err := rd.Read()
+	if err != nil {
+		t.Fatal("Can't read samples:", err)
+	}
+
+	want := []byte{1, 2, 3, 4, 4, 0, 0, 0, 0, 0, 0, 0}
+	if !bytes.Equal(record.RawSample, want) {
+		t.Log(record.RawSample)
+		t.Error("Sample doesn't match expected output")
+	}
+
+	if record.CPU < 0 {
+		t.Error("Record has invalid CPU number")
+	}
+}
+
+func outputSamplesProg(sampleSizes ...int) (*ebpf.Program, *ebpf.Map, error) {
+	const bpfFCurrentCPU = 0xffffffff
+
+	events, err := ebpf.NewMap(&ebpf.MapSpec{
+		Type: ebpf.PerfEventArray,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var maxSampleSize int
+	for _, sampleSize := range sampleSizes {
+		if sampleSize > maxSampleSize {
+			maxSampleSize = sampleSize
+		}
+	}
+
+	// Fill a buffer on the stack, and stash context somewhere
+	insns := asm.Instructions{
+		asm.LoadImm(asm.R0, 0x0102030404030201, asm.DWord),
+		asm.Mov.Reg(asm.R9, asm.R1),
+	}
+
+	bufDwords := (maxSampleSize / 8) + 1
+	for i := 0; i < bufDwords; i++ {
+		insns = append(insns,
+			asm.StoreMem(asm.RFP, int16(i+1)*-8, asm.R0, asm.DWord),
+		)
+	}
+
+	for _, sampleSize := range sampleSizes {
+		insns = append(insns,
+			asm.Mov.Reg(asm.R1, asm.R9),
+			asm.LoadMapPtr(asm.R2, events.FD()),
+			asm.LoadImm(asm.R3, bpfFCurrentCPU, asm.DWord),
+			asm.Mov.Reg(asm.R4, asm.RFP),
+			asm.Add.Imm(asm.R4, int32(bufDwords*-8)),
+			asm.Mov.Imm(asm.R5, int32(sampleSize)),
+			asm.FnPerfEventOutput.Call(),
+		)
+	}
+
+	insns = append(insns, asm.Return())
+
+	prog, err := ebpf.NewProgram(&ebpf.ProgramSpec{
+		License:      "GPL",
+		Type:         ebpf.XDP,
+		Instructions: insns,
+	})
+	if err != nil {
+		events.Close()
+		return nil, nil, err
+	}
+
+	return prog, events, nil
+}
+
+func mustOutputSamplesProg(tb testing.TB, sampleSizes ...int) (*ebpf.Program, *ebpf.Map) {
+	tb.Helper()
+
+	prog, events, err := outputSamplesProg(sampleSizes...)
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	return prog, events
+}
+
+func TestPerfReaderLostSample(t *testing.T) {
+	// To generate a lost sample perf record:
+	//
+	// 1. Fill the perf ring buffer almost completely, with the output_large program.
+	//    The buffer is sized in number of pages, which are architecture dependant.
+	//
+	// 2. Write an extra event that doesn't fit in the space remaining.
+	//
+	// 3. Write a smaller event that does fit, with output_single program.
+	//    Lost sample records are generated opportunistically, when the kernel
+	//    is writing an event and realizes that there were events lost previously.
```

Comment:
This is awesome.

I was wondering about another case which is basically the writing of the events over the barrier of the ringbuffer, where you would expect the first half of the event to be at the end of the buffer and the last half is at the start of the buffer.

You might imagine writing a test similar to this one, where first you almost fill the buffer, then consume all of the events, then write one more event across the end (/start) to loop back around and make sure this final event is correctly read out.

Anyway, that can be part of a subsquent follow-up.

Reply:
Yeah, good idea. There is a test for the wrap around behaviour of the ring reader, `TestRingBufferReader`. Have a look whether we need more.